README.md
# microbench

[![crates.io](https://img.shields.io/crates/v/microbench.svg)](https://crates.io/crates/microbench)
[![docs.rs](https://docs.rs/microbench/badge.svg)](https://docs.rs/microbench)
[![Travis CI](https://travis-ci.org/KyleMayes/microbench.svg?branch=master)](https://travis-ci.org/KyleMayes/microbench)

A micro-benchmarking library (inspired by [core_bench](https://github.com/janestreet/core_bench)).

Released under the Apache License 2.0.

Supported on Rust 1.31.0 and later.

**Note:** The `retain` function (used to prevent the optimizer from removing computations) may not
operate correctly or may have poor performance on the stable and beta channels of Rust. If you are
using a nightly release of Rust, enable the `nightly` crate feature to enable a better
implementation of this function.

## Overview

`microbench` uses linear regression to estimate the execution time of code segments. For
example, the following table might represent data collected by `microbench` about a code
segment.

| Iterations | Time (ns) |
|------------|-----------|
| 1          | 19        |
| 2          | 25        |
| 3          | 37        |
| 4          | 47        |
| 5          | 56        |

`microbench` of course takes many more than 5 samples and the number of iterations grows
geometrically rather than linearly, but the idea remains the same. After collecting data like
this, `microbench` uses ordinary least squares (OLS) linear regression to estimate the actual
execution time of the code segment. Using OLS with the above data would yield an estimated
execution time of `9.6` nanoseconds with a goodness of fit (R²) of `0.992`.

## Example

```rust
use microbench::{self, Options};

fn fibonacci_iterative(n: u64) -> u64 {
    let (mut x, mut y, mut z) = (0, 1, 1);
    for _ in 0..n { x = y; y = z; z = x + y; }
    x
}

fn fibonacci_recursive(n: u64) -> u64 {
    if n < 2 {
        n
    } else {
        fibonacci_recursive(n - 2) + fibonacci_recursive(n - 1)
    }
}

let options = Options::default();
microbench::bench(&options, "iterative_16", || fibonacci_iterative(16));
microbench::bench(&options, "recursive_16", || fibonacci_recursive(16));
```

Example output:

```console
iterative_16 (5.0s) ...                  281.733 ns/iter (0.998 R²)
recursive_16 (5.0s) ...                9_407.020 ns/iter (0.997 R²)
```


CHANGELOG.md
## [0.5.0] - 2019-04-03

### Fixed
- Updated `retain` implementation for latest nightly

## [0.4.0] - 2018-12-30

### Removed
- Removed `GeometricSequence` struct

### Changed
- Upgraded to Rust 2018
- Replaced `regression` function with `Model` struct
- Replaced `Statistics` trait with `Kahan` trait
- Replaced `Stopwatch::new` method with `Default` implementation

## [0.3.3] - 2018-08-18

### Removed
- Removed optional `clippy` dependency

## [0.3.2] - 2017-02-28

### Fixed
- Fixed integer overflow caused by benchmarking extremely fast functions

## [0.3.1] - 2017-01-29

### Added
- Added support for the `stable` and `beta` channels of Rust

### Changed
- Renamed `maximum` option to `time`

## [0.3.0] - 2016-12-28

### Added
- Added `memory` option
- Added specialized benchmarking functions (`bench_drop` and `bench_setup`)
- Added specialized measuring functions (`measure_drop` and `measure_setup`)

### Removed
- Removed `time` dependency
- Removed `GeometricSequence` struct
- Removed `Stopwatch` struct

### Changed
- Renamed `maximum` option to `time`

## [0.2.1] - 2016-12-26

### Changed
- Improved accuracy of results by using the Kahan summation algorithm

## [0.2.0] - 2016-12-25

### Added
- Added `alpha` field to `Anaylsis` struct

### Changed
- Improved accuracy of results by using the Kahan summation algorithm
- Improved formatting of printed results

## [0.1.0] - 2016-12-24
- Initial release


src/time.rs
// Copyright 2016 Kyle Mayes
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Time-related utilities.

use std::fmt;
use std::time::{Duration, Instant};

/// A number of nanoseconds.
#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Nanoseconds<T>(pub T);

impl fmt::Display for Nanoseconds<u64> {
    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        write!(formatter, "{:.1}s", self.0 as f64 / 1_000_000_000.0)
    }
}

impl From<Duration> for Nanoseconds<u64> {
    fn from(duration: Duration) -> Nanoseconds<u64> {
        let nanos = u64::from(duration.subsec_nanos());
        Nanoseconds((duration.as_secs() * 1_000_000_000) + nanos)
    }
}

/// A high-precision stopwatch.
#[derive(Clone, Copy, Debug)]
pub struct Stopwatch(Instant);

impl Stopwatch {
    /// Returns the number of nanoseconds that have elapsed since this stopwatch
    /// was last constructed or reset.
    pub fn elapsed(self) -> Nanoseconds<u64> {
        (Instant::now() - self.0).into()
    }
}

impl Default for Stopwatch {
    fn default() -> Self {
        Stopwatch(Instant::now())
    }
}


src/utility.rs
// Copyright 2016 Kyle Mayes
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Miscellaneous utilities.

/// Generates unique values from a geometric sequence.
#[derive(Copy, Clone, Debug)]
pub struct GeometricSequence {
    current: f64,
    factor: f64,
}

impl GeometricSequence {
    /// Constructs a new `GeometricSequence`.
    pub fn new(start: u64, factor: f64) -> Self {
        GeometricSequence { current: start as f64, factor }
    }
}

impl Iterator for GeometricSequence {
    type Item = u64;

    fn next(&mut self) -> Option<Self::Item> {
        let start = self.current as u64;
        while self.current as u64 == start { self.current *= self.factor; }
        Some(start)
    }
}

/// A function that prevents the optimizer from eliminating the supplied value.
#[cfg(feature="nightly")]
pub fn black_box<T>(dummy: T) -> T {
    use std::hint;
    hint::black_box(dummy)
}

/// A function that prevents the optimizer from eliminating the supplied value.
#[cfg(not(feature="nightly"))]
pub fn black_box<T>(dummy: T) -> T {
    use std::mem;
    use std::ptr;
    unsafe {
        let value = ptr::read_volatile(&dummy);
        mem::forget(dummy);
        value
    }
}

/// Returns the supplied floating-point number formatted with the supplied
/// precision and thousands separator.
pub fn format_number(number: f64, precision: usize, separator: char) -> String {
    let mut integral = String::new();

    let mut counter = 0;
    for digit in (number as u64).to_string().chars().rev() {
        if counter == 3 {
            integral.insert(0, separator);
            counter = 0;
        }

        counter += 1;
        integral.insert(0, digit);
    }

    let fractional = format!("{:.*}", precision, number.fract());
    format!("{}.{}", integral, &fractional[2..])
}


src/statistics.rs
// Copyright 2016 Kyle Mayes
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Statistics-related utilities.

use std::iter::{FromIterator};

/// A collection of floating-point numbers that supports Kahan summation.
pub trait Kahan {
    /// Returns the mean of the numbers in this collection using the Kahan
    /// summation algorithm.
    fn kahan_mean(self) -> f64;
    /// Returns the sum of the numbers in this collection using the Kahan
    /// summation algorithm.
    fn kahan_sum(self) -> f64;
}

impl<I> Kahan for I where I: ExactSizeIterator<Item=f64> {
    fn kahan_mean(self) -> f64 {
        let size = self.len() as f64;
        self.kahan_sum() / size
    }

    fn kahan_sum(self) -> f64 {
        self.fold((0.0, 0.0), |(sum, correction), f| {
            let y = f - correction;
            let t = sum + y;
            (t, (t - sum) - y)
        }).0
    }
}

/// A simple linear regression model.
#[derive(Copy, Clone, Debug)]
pub struct Model {
    /// The y-intercept of the model function.
    pub alpha: f64,
    /// The slope of the model function.
    pub beta: f64,
    /// The goodness of fit of the model function.
    pub r2: f64,
}

impl Model {
    /// Returns a new model for the supplied data using OLS linear regression.
    fn new(data: &[(f64, f64)]) -> Self {
        let xmean = data.iter().map(|d| d.0).kahan_mean();
        let ymean = data.iter().map(|d| d.1).kahan_mean();

        // OLS linear regression.
        let numerator = data.iter().map(|m| (m.0 - xmean) * (m.1 - ymean)).kahan_sum();
        let denominator = data.iter().map(|m| (m.0 - xmean).powf(2.0)).kahan_sum();
        let beta = numerator / denominator;
        let alpha = ymean - (beta * xmean);
        let estimator = |x| (beta * x) + alpha;

        // OLS goodness of fit.
        let numerator = data.iter().map(|m| (estimator(m.0) - ymean).powf(2.0)).kahan_sum();
        let denominator = data.iter().map(|m| (m.1 - ymean).powf(2.0)).kahan_sum();
        let r2 = numerator / denominator;

        Self { alpha, beta, r2 }
    }
}

impl FromIterator<(f64, f64)> for Model {
    fn from_iter<I>(iter: I) -> Self where I: IntoIterator<Item=(f64, f64)> {
        Model::new(&iter.into_iter().collect::<Vec<_>>())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_kahan() {
        let numbers: &[f64] = &[10000.0, 3.14159, 2.71828];
        assert_eq!(numbers.iter().cloned().kahan_sum(), 10005.85987);
    }

    #[test]
    fn test_model() {
        let data: &[(f64, f64)] = &[
            (1.47, 52.21), (1.50, 53.12), (1.52, 54.48),
            (1.55, 55.84), (1.57, 57.20), (1.60, 58.57),
            (1.63, 59.93), (1.65, 61.29), (1.68, 63.11),
            (1.70, 64.47), (1.73, 66.28), (1.75, 68.10),
            (1.78, 69.92), (1.80, 72.19), (1.83, 74.46),
        ];

        let model = data.iter().cloned().collect::<Model>();
        assert_eq!(model.alpha, -39.06195591884393);
        assert_eq!(model.beta, 61.27218654211062);
        assert_eq!(model.r2, 0.989196922445796);
    }
}


src/lib.rs
// Copyright 2016 Kyle Mayes
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! A micro-benchmarking library.
//!
//! # Overview
//!
//! `microbench` uses linear regression to estimate the execution time of code
//! segments. For example, the following table might represent data collected by
//! `microbench` about a code segment.
//!
//! | Iterations | Time (ns) |
//! |------------|-----------|
//! | 1          | 19        |
//! | 2          | 25        |
//! | 3          | 37        |
//! | 4          | 47        |
//! | 5          | 56        |
//!
//! `microbench` of course takes many more than 5 samples and the number of
//! iterations grows geometrically rather than linearly, but the idea remains
//! the same. After collecting data like this, `microbench` uses ordinary least
//! squares (OLS) linear regression to estimate the actual execution time of the
//! code segment. Using OLS with the above data would yield an estimated
//! execution time of `9.6` nanoseconds with a goodness of fit (R²) of `0.992`.
//!
//! # Example
//!
//! ```
//! use microbench::{self, Options};
//!
//! fn fibonacci_iterative(n: u64) -> u64 {
//!     let (mut x, mut y, mut z) = (0, 1, 1);
//!     for _ in 0..n { x = y; y = z; z = x + y; }
//!     x
//! }
//!
//! fn fibonacci_recursive(n: u64) -> u64 {
//!     if n < 2 {
//!         n
//!     } else {
//!         fibonacci_recursive(n - 2) + fibonacci_recursive(n - 1)
//!     }
//! }
//!
//! let options = Options::default();
//! microbench::bench(&options, "iterative_16", || fibonacci_iterative(16));
//! microbench::bench(&options, "recursive_16", || fibonacci_recursive(16));
//! ```
//!
//! Example output:
//!
//! ```console
//! iterative_16 (5.0s) ...                  281.733 ns/iter (0.998 R²)
//! recursive_16 (5.0s) ...                9_407.020 ns/iter (0.997 R²)
//! ```

#![cfg_attr(feature="nightly", feature(test))]

#![warn(missing_copy_implementations, missing_debug_implementations, missing_docs)]

mod utility;
pub mod statistics;
pub mod time;

use std::cmp;
use std::mem;
use std::time::{Duration};

use crate::statistics::{Model};
use crate::time::{Nanoseconds, Stopwatch};
use crate::utility::{GeometricSequence, black_box, format_number};

/// The maximum number of benchmark sample iterations.
const ITERATIONS: u64 = 1_000_000_000_000_000;

/// A number of bytes.
#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct Bytes(pub u64);

impl Bytes {
    /// Returns the number of bytes in the supplied number of kibibytes (2¹⁰ bytes).
    pub fn kibibytes(kibibytes: u64) -> Self {
        Bytes(kibibytes * 1024)
    }

    /// Returns the number of bytes in the supplied number of mebibytes (2²⁰ bytes).
    pub fn mebibytes(mebibytes: u64) -> Self {
        Bytes(mebibytes * 1024 * 1024)
    }

    /// Returns the number of bytes in the supplied number of gibibytes (2³⁰ bytes).
    pub fn gibibytes(gibibytes: u64) -> Self {
        Bytes(gibibytes * 1024 * 1024 * 1024)
    }
}

/// A set of benchmarking options.
#[derive(Copy, Clone, Debug)]
pub struct Options {
    factor: f64,
    memory: Bytes,
    time: Nanoseconds<u64>,
}

impl Options {
    /// Sets the geometric growth factor for benchmark sample iterations.
    ///
    /// **Default:** `1.01`
    pub fn factor(mut self, factor: f64) -> Self {
        self.factor = factor;
        self
    }

    /// Sets the maximum amount of memory benchmarks will allocate.
    ///
    /// **Default:** `Bytes::mebibytes(512)`
    pub fn memory(mut self, memory: Bytes) -> Self {
        self.memory = memory;
        self
    }

    /// Sets the maximum amount of time benchmarks will run for.
    ///
    /// **Default:** `Duration::new(5, 0)`
    pub fn time(mut self, time: Duration) -> Self {
        self.time = time.into();
        self
    }
}

impl Default for Options {
    fn default() -> Self {
        let factor = 1.01;
        let memory = Bytes::mebibytes(512);
        let time = Duration::new(5, 0).into();
        Options { factor, memory, time }
    }
}

/// A sample of the execution time of a function.
#[derive(Copy, Clone, Debug)]
pub struct Sample {
    /// The number of times the function was executed.
    pub iterations: u64,
    /// The number of nanoseconds that elapsed while executing the function.
    pub elapsed: Nanoseconds<u64>,
}

/// A statistical analysis of a set of execution time samples.
#[derive(Copy, Clone, Debug)]
pub struct Analysis {
    /// The y-intercept of the simple linear regression model function.
    pub alpha: Nanoseconds<f64>,
    /// The slope of the simple linear regression model function.
    pub beta: Nanoseconds<f64>,
    /// The goodness of fit of the simple linear regression model function.
    pub r2: f64,
}

impl Analysis {
    /// Returns a new analysis for the supplied samples.
    fn new(samples: &[Sample]) -> Self {
        let Model { alpha, beta, r2 } = samples.iter()
            .map(|m| (m.iterations as f64, m.elapsed.0 as f64))
            .collect::<Model>();
        Self { alpha: Nanoseconds(alpha), beta: Nanoseconds(beta), r2 }
    }
}

/// Benchmarks the supplied function and prints the results.
pub fn bench<T>(options: &Options, name: &str, f: impl FnMut() -> T) {
    bench_impl(name, move || measure(options, f));
}

/// Benchmarks the supplied function ignoring drop time and prints the results.
///
/// See [`measure_drop`](fn.measure_drop.html) for more information.
pub fn bench_drop<T>(options: &Options, name: &str, f: impl FnMut() -> T) {
    bench_impl(name, move || measure_drop(options, f));
}

/// Benchmarks the supplied function ignoring setup time and prints the results.
///
/// See [`measure_setup`](fn.measure_setup.html) for more information.
pub fn bench_setup<I, T>(
    options: &Options,
    name: &str,
    setup: impl FnMut() -> I,
    f: impl FnMut(I) -> T,
) {
    bench_impl(name, move || measure_setup(options, setup, f));
}

/// Measures the execution time of the supplied function.
pub fn measure<T>(
    options: &Options, mut f: impl FnMut() -> T
) -> Vec<Sample> {
    measure_impl(options, |iterations| {
        let stopwatch = Stopwatch::default();
        for _ in 0..iterations { retain(f()); }
        Some(stopwatch.elapsed())
    })
}

/// Measures the execution time of the supplied function ignoring drop time.
///
/// This function does not include the time it takes to drop the values returned
/// by the supplied function in the measurements. This can be useful when you
/// want to exclude the running time of a slow implementation of `Drop` from
/// your benchmark. However, it should be noted that this function introduces a
/// very small amount of overhead which will be reflected in the measurements
/// (typically of the order of a few nanoseconds).
///
/// **Warning:** This function can potentially allocate very large amounts of
/// memory. The `memory` option controls the maximum amount of memory this
/// function is allowed to allocate.
pub fn measure_drop<T>(
    options: &Options, mut f: impl FnMut() -> T
) -> Vec<Sample> {
    measure_impl(options, |iterations| {
        let size = cmp::max(1, mem::size_of::<T>() as u64);
        if options.memory < Bytes(iterations * size) {
            return None;
        }

        let mut outputs = Vec::with_capacity(iterations as usize);
        let stopwatch = Stopwatch::default();
        for _ in 0..iterations { outputs.push(f()); }
        let elapsed = stopwatch.elapsed();
        mem::drop(outputs);
        Some(elapsed)
    })
}

/// Measures the execution time of the supplied function ignoring setup time.
///
/// This function does not include the time it takes to execute the setup
/// function in the measurements. This can be useful when you want to exclude
/// the running time of some non-trivial setup which is needed for every
/// execution of the supplied function. However, it should be noted that this
/// function introduces a very small amount of overhead which will be reflected
/// in the measurements (typically of the order of a few nanoseconds).
///
/// **Warning:** This function can potentially allocate very large amounts of
/// memory. The `memory` option controls the maximum amount of memory this
/// function is allowed to allocate.
pub fn measure_setup<I, T>(
    options: &Options,
    mut setup: impl FnMut() -> I,
    mut f: impl FnMut(I) -> T,
) -> Vec<Sample> {
    measure_impl(options, |iterations| {
        let size = cmp::max(1, mem::size_of::<I>() as u64);
        if options.memory < Bytes(iterations * size) {
            return None;
        }

        let inputs = retain((0..iterations).map(|_| setup()).collect::<Vec<_>>());
        let stopwatch = Stopwatch::default();
        for input in inputs { retain(f(input)); }
        Some(stopwatch.elapsed())
    })
}

/// A function that prevents the optimizer from eliminating the supplied value.
///
/// This function may not operate correctly or may have poor performance on the
/// stable and beta channels of Rust. If you are using a nightly release of
/// Rust, enable the `nightly` crate feature to enable a superior implementation
/// of this function.
pub fn retain<T>(value: T) -> T {
    black_box(value)
}

/// Prints an analysis of the samples produced by the supplied function.
fn bench_impl(name: &str, f: impl FnOnce() -> Vec<Sample>) {
    let stopwatch = Stopwatch::default();
    let samples = f();
    let elapsed = stopwatch.elapsed();
    let analysis = Analysis::new(&samples);

    let prefix = format!("{} ({}) ...", name, elapsed);
    if samples.len() < 2 || analysis.beta.0 < 0.0 {
        println!("{:<32} {:>15}", prefix, "           not enough samples");
    } else {
        let beta = format_number(analysis.beta.0, 3, '_');
        println!("{:<32} {:>15} ns/iter ({:.3} R²)", prefix, beta, analysis.r2);
    }
}

/// Collects samples produced by the supplied function.
fn measure_impl(
    options: &Options, mut f: impl FnMut(u64) -> Option<Nanoseconds<u64>>
) -> Vec<Sample> {
    let stopwatch = Stopwatch::default();
    GeometricSequence::new(1, options.factor)
        .take_while(|i| *i <= ITERATIONS && stopwatch.elapsed() < options.time)
        .filter_map(|i| Some(Sample { iterations: i, elapsed: f(i)? }))
        .collect()
}
